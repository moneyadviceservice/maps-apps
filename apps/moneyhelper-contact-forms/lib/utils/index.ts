// export client-side safe utilities only
export * from './getCurrentStep';
export * from './getFieldError';
export * from './getFlowSteps';
export * from './getSessionId';
export * from './preparePayload';
export * from './safeT';
