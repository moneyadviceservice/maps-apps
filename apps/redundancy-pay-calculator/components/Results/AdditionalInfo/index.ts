import { AdditionalInfo } from './AdditionalInfo';

export { AdditionalInfo };
