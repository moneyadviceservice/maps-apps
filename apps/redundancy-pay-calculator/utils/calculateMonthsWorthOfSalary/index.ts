import { calculateMonthsWorthOfSalary } from './calculateMonthsWorthOfSalary';

export { calculateMonthsWorthOfSalary };
