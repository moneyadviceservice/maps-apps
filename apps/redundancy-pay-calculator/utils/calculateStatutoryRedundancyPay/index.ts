import {
  calculateStatutoryRedundancyPay,
  calculateYearlyPay,
  getDateFromDMY,
  getDateFromMY,
  Salary,
  StatutoryRedundancyPayInput,
  StatutoryRedundancyPayResult,
} from './calculateStatutoryRedundancyPay';

export {
  calculateStatutoryRedundancyPay,
  calculateYearlyPay,
  getDateFromDMY,
  getDateFromMY,
};
export type {
  Salary,
  StatutoryRedundancyPayInput,
  StatutoryRedundancyPayResult,
};
