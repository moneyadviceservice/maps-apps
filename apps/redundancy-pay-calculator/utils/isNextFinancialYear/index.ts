import { isNextFinancialYear } from './isNextFinancialYear';

export { isNextFinancialYear };
