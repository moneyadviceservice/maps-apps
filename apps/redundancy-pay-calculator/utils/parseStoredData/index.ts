import {
  ContractualRedundancyProvided,
  Country,
  formatNumber,
  ParsedData,
  parseStoredData,
} from './parseStoredData';

export {
  ContractualRedundancyProvided,
  Country,
  formatNumber,
  parseStoredData,
};
export type { ParsedData };
