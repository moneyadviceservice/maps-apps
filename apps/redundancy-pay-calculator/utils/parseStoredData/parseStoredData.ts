import { DataFromQuery } from '@maps-react/utils/pageFilter';

import { Salary } from '../../utils/calculateStatutoryRedundancyPay';

export enum Country {
  England,
  Scotland,
  Wales,
  NorthernIreland,
}

export enum ContractualRedundancyProvided {
  Yes,
  No,
  Unknown,
}

type ContractualRedundancyAmount =
  | number
  | ContractualRedundancyProvided.Unknown;

export type ContractualRedundancy = {
  provided: ContractualRedundancyProvided;
  amount: number | ContractualRedundancyProvided.Unknown;
};

export type ParsedData = {
  dateOfBirth: string | undefined;
  salary: Salary | undefined;
  jobStart: string | undefined;
  jobEnd: string | undefined;
  country: Country | undefined;
  contractualRedundancy: ContractualRedundancy | undefined;
};

/**
 * Parses the stored data from a query and maps it to a structured format.
 *
 * @param storedData - The raw data retrieved from a query, containing key-value pairs.
 * @returns An object containing parsed and structured data, including:
 * - `dateOfBirth`: The date of birth extracted from the query data.
 * - `salary`: The parsed salary value.
 * - `jobStart`: The start date of the job.
 * - `jobEnd`: The end date of the job.
 * - `country`: The parsed country information.
 * - `contractualRedundancy`: The parsed contractual redundancy details.
 */
export const parseStoredData = (storedData: DataFromQuery): ParsedData => {
  return {
    dateOfBirth: storedData['q-2'],
    salary: parseSalary(storedData['q-5']),
    jobStart: storedData['q-4'],
    jobEnd: storedData['q-3'],
    country: parseCountry(storedData['q-1']),
    contractualRedundancy: parseContractualRedundancy(
      storedData['q-6'],
      storedData['q-7'],
    ),
  };
};

/**
 * Parses the provided string to determine the corresponding country.
 *
 * @param q1 - A string representing the country identifier.
 * @returns The parsed `Country` if the input is valid, or `undefined` if the input is empty or invalid.
 */
const parseCountry = (q1: string): Country | undefined => {
  if (!q1) return undefined;

  return Number(q1);
};

/**
 * Parses a salary string into a `Salary` object.
 *
 * The input string `q5` is expected to have a format where the salary amount
 * and frequency are separated by a comma. For example: "50,000,12".
 *
 * - The portion before the last comma is treated as the salary amount.
 * - The portion after the last comma is treated as the frequency.
 *
 * @param q5 - The salary string to parse. If the string is empty or undefined, the function returns `undefined`.
 * @returns A `Salary` object containing the `amount` and `frequency`, or `undefined` if the input is invalid.
 */
const parseSalary = (q5: string): Salary | undefined => {
  if (!q5) return undefined;

  const lastCommaIndex = q5.lastIndexOf(',');
  const amount = Number(q5.substring(0, lastCommaIndex).replaceAll(',', ''));
  const frequency = Number(q5.substring(lastCommaIndex + 1));

  return { amount, frequency };
};

/**
 * Parses the contractual redundancy information based on the provided inputs.
 *
 * @param q6 - A string representing whether contractual redundancy is provided.
 *             Expected values:
 *             - '1' for Yes
 *             - '2' for "I don't know"
 *             - Any other value is treated as No
 * @param q7 - A string representing the amount of contractual redundancy,
 *             formatted with commas (e.g., "1,000"). The last segment after
 *             the last comma indicates whether the amount is known:
 *             - '1' indicates the amount is unknown
 *             - Any other value indicates the amount is known
 *
 * @returns An object of type `ContractualRedundancy` containing:
 *          - `provided`: Indicates whether contractual redundancy is provided
 *                        (Yes, No, or Unknown).
 *          - `amount`: The known amount of contractual redundancy, or 0 if unknown.
 *          Returns `undefined` if the input data is insufficient or invalid.
 */
export const parseContractualRedundancy = (
  q6: string,
  q7: string,
): ContractualRedundancy | undefined => {
  if (!q6) return undefined;

  let provided = Number(q6);

  let amount: ContractualRedundancyAmount = 0;

  const q6IDontKnow = q6 === '2';
  if (q6IDontKnow) {
    provided = ContractualRedundancyProvided.No;
  }

  if (provided === ContractualRedundancyProvided.Yes) {
    if (!q7) return undefined;

    const lastCommaIndex = q7.lastIndexOf(',');
    const isKnown = Number(q7.substring(lastCommaIndex + 1)) !== 1;
    const knownAmount = Number(q7.substring(0).replaceAll(',', ''));

    if (isKnown) {
      amount = knownAmount;
    } else {
      provided = ContractualRedundancyProvided.Unknown;
    }
  }

  return {
    provided,
    amount,
  };
};

/**
 * Formats a number into a localized string representation using the 'en-GB' locale.
 * The formatted string includes up to two decimal places.
 *
 * @param number - The number to be formatted.
 * @returns The formatted number as a string.
 */
export const formatNumber = (number: number): string =>
  number.toLocaleString('en-GB', { maximumFractionDigits: 2 });
