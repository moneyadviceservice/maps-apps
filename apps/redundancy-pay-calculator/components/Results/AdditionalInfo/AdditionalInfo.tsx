import { Paragraph } from '@maps-react/common/index';
import { useTranslation } from '@maps-react/hooks/useTranslation';

import copy from '../../../data/form-content/text/results';
import {
  getDateFromMY,
  StatutoryRedundancyPayResult,
} from '../../../utils/calculateStatutoryRedundancyPay';
import { isNextFinancialYear } from '../../../utils/isNextFinancialYear';
import {
  ContractualRedundancyProvided,
  ParsedData,
} from '../../../utils/parseStoredData';

type Props = {
  parsedData: ParsedData;
  statutoryRedundancyPay: StatutoryRedundancyPayResult;
  yearsWorked: number;
};

export const AdditionalInfo = ({
  parsedData,
  statutoryRedundancyPay,
  yearsWorked,
}: Props) => {
  const { z } = useTranslation();

  if (
    undefined === parsedData.jobEnd ||
    undefined === parsedData.contractualRedundancy
  )
    return;

  const jobEndDate = getDateFromMY(parsedData.jobEnd);

  const isEndDateInNextFinancialYear = isNextFinancialYear(jobEndDate) === 1;

  const isContractualLessThanStatutory =
    parsedData.contractualRedundancy.amount < statutoryRedundancyPay.amount;

  const hasContractualPay =
    parsedData.contractualRedundancy.provided ===
    ContractualRedundancyProvided.Yes;

  const noContractualPay =
    parsedData.contractualRedundancy.provided ===
    ContractualRedundancyProvided.No;

  const contractualAmountUnknown =
    parsedData.contractualRedundancy.provided ===
    ContractualRedundancyProvided.Unknown;

  return (
    <div className="pt-2 pb-6">
      {yearsWorked >= 2 && (
        <>
          {isEndDateInNextFinancialYear && (
            <>
              {(noContractualPay ||
                contractualAmountUnknown ||
                (hasContractualPay &&
                  !contractualAmountUnknown &&
                  isContractualLessThanStatutory)) && (
                <Paragraph>
                  {z(copy.additionalInfo.nextFinancialYear)}{' '}
                </Paragraph>
              )}
            </>
          )}
          {(contractualAmountUnknown ||
            (hasContractualPay && isContractualLessThanStatutory)) && (
            <Paragraph>
              {z(copy.additionalInfo.contractualUnknownOrLessThanStatutory)}
            </Paragraph>
          )}
        </>
      )}
      {yearsWorked < 2 && contractualAmountUnknown && (
        <Paragraph>
          {z(copy.additionalInfo.contractualUnknownOrLessThanStatutory)}
        </Paragraph>
      )}
    </div>
  );
};
