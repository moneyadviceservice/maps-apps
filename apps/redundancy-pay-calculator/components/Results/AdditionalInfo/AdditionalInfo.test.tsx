import { render } from '@testing-library/react';
import { AdditionalInfo } from './AdditionalInfo';

import '@testing-library/jest-dom/extend-expect';

import { ContractualRedundancyProvided } from '../../../utils/parseStoredData';
import copy from 'data/form-content/text/results';

jest.mock('@maps-react/hooks/useTranslation', () => ({
  useTranslation: () => ({
    z: jest.fn((obj) => obj.en),
  }),
}));

describe('Redundancy pay calculator', () => {
  describe('AdditionalInfo Component', () => {
    it.each`
      dateOfBirth   | salary                             | jobStart    | jobEnd      | country | contractualRedundancy                                                                             | yearsWorked | statutoryRedundancyPay
      ${'4-5-1980'} | ${{ amount: 15000, frequency: 0 }} | ${'3-2020'} | ${'2-2025'} | ${0}    | ${{ provided: ContractualRedundancyProvided.Unknown, amount: 0 }}                                 | ${5}        | ${{ amount: 1587, entitlementWeeks: 5.5 }}
      ${'4-5-1980'} | ${{ amount: 15000, frequency: 0 }} | ${'3-2020'} | ${'2-2025'} | ${0}    | ${{ provided: ContractualRedundancyProvided.Yes, amount: 2000 }}                                  | ${5}        | ${{ amount: 1587, entitlementWeeks: 5.5 }}
      ${'4-5-1980'} | ${{ amount: 15000, frequency: 0 }} | ${'3-2020'} | ${'2-2025'} | ${0}    | ${{ provided: ContractualRedundancyProvided.Yes, amount: 1000 }}                                  | ${5}        | ${{ amount: 1587, entitlementWeeks: 5.5 }}
      ${'4-5-1980'} | ${{ amount: 15000, frequency: 0 }} | ${'3-2020'} | ${'2-2025'} | ${0}    | ${{ provided: ContractualRedundancyProvided.Yes, amount: ContractualRedundancyProvided.Unknown }} | ${5}        | ${{ amount: 1587, entitlementWeeks: 5.5 }}
      ${'4-5-1980'} | ${{ amount: 15000, frequency: 0 }} | ${'3-2020'} | ${'4-2025'} | ${0}    | ${{ provided: ContractualRedundancyProvided.No, amount: 0 }}                                      | ${5}        | ${{ amount: 1587, entitlementWeeks: 5.5 }}
      ${'4-5-1980'} | ${{ amount: 15000, frequency: 0 }} | ${'3-2024'} | ${'2-2025'} | ${0}    | ${{ provided: ContractualRedundancyProvided.Yes, amount: ContractualRedundancyProvided.Unknown }} | ${1}        | ${{ amount: 1587, entitlementWeeks: 5.5 }}
      ${'4-5-1980'} | ${{ amount: 15000, frequency: 0 }} | ${'3-2024'} | ${'4-2025'} | ${0}    | ${{ provided: ContractualRedundancyProvided.Yes, amount: 2000 }}                                  | ${2}        | ${{ amount: 1587, entitlementWeeks: 5.5 }}
      ${'4-5-1980'} | ${{ amount: 15000, frequency: 0 }} | ${'3-2020'} | ${'3-2023'} | ${0}    | ${{ provided: ContractualRedundancyProvided.Yes, amount: 1500 }}                                  | ${2}        | ${{ amount: 1587, entitlementWeeks: 5.5 }}
    `(
      'renders the component with provided props',
      ({
        dateOfBirth,
        salary,
        jobStart,
        jobEnd,
        country,
        contractualRedundancy,
        yearsWorked,
        statutoryRedundancyPay,
      }) => {
        const parsedData = {
          dateOfBirth,
          salary,
          jobStart,
          jobEnd,
          country,
          contractualRedundancy: contractualRedundancy,
        };

        const { container } = render(
          <AdditionalInfo
            parsedData={parsedData}
            statutoryRedundancyPay={statutoryRedundancyPay}
            yearsWorked={yearsWorked}
          />,
        );

        expect(container.firstChild).toMatchSnapshot();
      },
    );

    it('renders correctly for yearsWorked < 2 and unknown contractual redundancy', () => {
      const parsedData = {
        dateOfBirth: '4-5-1980',
        salary: { amount: 15000, frequency: 0 },
        jobStart: '3-2024',
        jobEnd: '4-2025',
        country: 0,
        contractualRedundancy: {
          provided: ContractualRedundancyProvided.Unknown,
          amount: 0,
        },
      };
      const statutoryRedundancyPay = { amount: 1587, entitlementWeeks: 5.5 };
      const yearsWorked = 1;

      const { container } = render(
        <AdditionalInfo
          parsedData={parsedData}
          statutoryRedundancyPay={statutoryRedundancyPay}
          yearsWorked={yearsWorked}
        />,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('renders correctly when statutory redundancy pay is higher than contractual redundancy', () => {
      const parsedData = {
        dateOfBirth: '4-5-1980',
        salary: { amount: 15000, frequency: 0 },
        jobStart: '3-2020',
        jobEnd: '2-2025',
        country: 0,
        contractualRedundancy: {
          provided: ContractualRedundancyProvided.Yes,
          amount: 1000,
        },
      };
      const statutoryRedundancyPay = { amount: 1587, entitlementWeeks: 5.5 };
      const yearsWorked = 5;

      const { container } = render(
        <AdditionalInfo
          parsedData={parsedData}
          statutoryRedundancyPay={statutoryRedundancyPay}
          yearsWorked={yearsWorked}
        />,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  it('returns early when jobEnd or contractualRedundancy is undefined', () => {
    const baseData = {
      dateOfBirth: '4-5-1980',
      salary: { amount: 15000, frequency: 0 },
      jobStart: '3-2020',
      country: 0,
    };

    const testCases = [
      {
        description: 'jobEnd is undefined',
        parsedData: {
          ...baseData,
          jobEnd: undefined,
          contractualRedundancy: {
            provided: ContractualRedundancyProvided.Unknown,
            amount: 0,
          },
        },
      },
      {
        description: 'contractualRedundancy is undefined',
        parsedData: {
          ...baseData,
          jobEnd: '2-2025',
          contractualRedundancy: undefined,
        },
      },
    ];

    const statutoryRedundancyPay = { amount: 1587, entitlementWeeks: 5.5 };
    const yearsWorked = 5;

    testCases.forEach(({ parsedData }) => {
      const { container } = render(
        <AdditionalInfo
          parsedData={parsedData}
          statutoryRedundancyPay={statutoryRedundancyPay}
          yearsWorked={yearsWorked}
        />,
      );

      expect(container.firstChild).toBeNull();
    });
  });

  it('renders additional info for job ending in next financial year with no contractual pay', () => {
    const parsedData = {
      dateOfBirth: '4-5-1980',
      salary: { amount: 15000, frequency: 0 },
      jobStart: '3-2020',
      jobEnd: '4-2026',
      country: 0,
      contractualRedundancy: {
        provided: ContractualRedundancyProvided.No,
        amount: 0,
      },
    };
    const statutoryRedundancyPay = { amount: 1587, entitlementWeeks: 5.5 };
    const yearsWorked = 5;

    const { getByText } = render(
      <AdditionalInfo
        parsedData={parsedData}
        statutoryRedundancyPay={statutoryRedundancyPay}
        yearsWorked={yearsWorked}
      />,
    );

    expect(
      getByText(copy.additionalInfo.nextFinancialYear.en),
    ).toBeInTheDocument();
  });
});
