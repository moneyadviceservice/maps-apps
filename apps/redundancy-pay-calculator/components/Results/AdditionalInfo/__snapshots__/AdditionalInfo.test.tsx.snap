// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redundancy pay calculator AdditionalInfo Component renders correctly for yearsWorked < 2 and unknown contractual redundancy 1`] = `
<div
  class="pt-2 pb-6"
>
  <p
    class="mb-4"
    data-testid="paragraph"
  >
    Check your employment contract or staff handbook to find out about your contractual pay. You can also ask your employer for the amount.
  </p>
</div>
`;

exports[`Redundancy pay calculator AdditionalInfo Component renders correctly when statutory redundancy pay is higher than contractual redundancy 1`] = `
<div
  class="pt-2 pb-6"
>
  <p
    class="mb-4"
    data-testid="paragraph"
  >
    Check your employment contract or staff handbook to find out about your contractual pay. You can also ask your employer for the amount.
  </p>
</div>
`;

exports[`Redundancy pay calculator AdditionalInfo Component renders the component with provided props 1`] = `
<div
  class="pt-2 pb-6"
>
  <p
    class="mb-4"
    data-testid="paragraph"
  >
    Check your employment contract or staff handbook to find out about your contractual pay. You can also ask your employer for the amount.
  </p>
</div>
`;

exports[`Redundancy pay calculator AdditionalInfo Component renders the component with provided props 2`] = `
<div
  class="pt-2 pb-6"
/>
`;

exports[`Redundancy pay calculator AdditionalInfo Component renders the component with provided props 3`] = `
<div
  class="pt-2 pb-6"
>
  <p
    class="mb-4"
    data-testid="paragraph"
  >
    Check your employment contract or staff handbook to find out about your contractual pay. You can also ask your employer for the amount.
  </p>
</div>
`;

exports[`Redundancy pay calculator AdditionalInfo Component renders the component with provided props 4`] = `
<div
  class="pt-2 pb-6"
>
  <p
    class="mb-4"
    data-testid="paragraph"
  >
    Check your employment contract or staff handbook to find out about your contractual pay. You can also ask your employer for the amount.
  </p>
</div>
`;

exports[`Redundancy pay calculator AdditionalInfo Component renders the component with provided props 5`] = `
<div
  class="pt-2 pb-6"
/>
`;

exports[`Redundancy pay calculator AdditionalInfo Component renders the component with provided props 6`] = `
<div
  class="pt-2 pb-6"
/>
`;

exports[`Redundancy pay calculator AdditionalInfo Component renders the component with provided props 7`] = `
<div
  class="pt-2 pb-6"
/>
`;

exports[`Redundancy pay calculator AdditionalInfo Component renders the component with provided props 8`] = `
<div
  class="pt-2 pb-6"
>
  <p
    class="mb-4"
    data-testid="paragraph"
  >
    Check your employment contract or staff handbook to find out about your contractual pay. You can also ask your employer for the amount.
  </p>
</div>
`;
